### PR TITLE
Consumer API: Terminate a Relationship

### PR DESCRIPTION
**Pull request by NikolaVetnic**
# Readiness checklist

-   [X] I added/updated unit tests.
-   [ ] I added/updated integration tests.
-   [X] I ensured that the PR title is good enough for the changelog.
-   [X] I labeled the PR.

# Description

Relationship can now be terminated via `PUT /Relationships/{id}/Terminate` endpoint, after which the status of the relationship is set to `Terminated`, an appropriate `RelationshipAuditLogEntry` is created, and a `RelationshipStatusChangedIntegrationEvent` integration event is triggered. 

After termination the participating identities cannot send messages to each another anymore. As long as a terminated relationship exists it is not possible to create a new one. Termination is only possible while the relationship is in active status.

### CODE REVIEW

**Comment by NikolaVetnic** [comment]
This may be nitpicking but I'll point it out in any case.

Regarding `Modules.Relationships.Application.Tests.TestHelpers.TestData` class and its `CreatePendingRelationship` method... There is an optional `IdentityAddress? to = null` parameter which is then not "processed" in the same way as `from` is:
```csharp
        from ??= TestDataGenerator.CreateRandomIdentityAddress();
```

This parameter is used to create a `RelationshipTemplate` thusly: `CreateRelationshipTemplate(createdBy: to)`. Even though the `createdBy` property is not nullable and should therefore not be `null`, it's possible to pass in `null` due to the EF Core constructor which does the `null` suppression with the `!` operator.

I would suggest doing the same I did for the `CreateActiveRelationship` method:
```csharp
        from ??= TestDataGenerator.CreateRandomIdentityAddress();
        to ??= TestDataGenerator.CreateRandomIdentityAddress();
```

File: Modules/Messages/src/Messages.Infrastructure/Persistence/Database/Repository/RelationshipsRepository.cs [hunk 1]
```diff
@@ -22,4 +23,12 @@ public RelationshipsRepository(MessagesDbContext dbContext)
             .Select(r => r.Id)
             .FirstOrDefaultAsync();
     }
+
+    public Task<Relationship?> FindRelationship(RelationshipId id, CancellationToken cancellationToken, bool track = false)
+    {
+        return _dbContext.Relationships
+            .AsNoTracking()
+            .Where(r => r.Id == id)
+            .FirstOrDefaultAsync(cancellationToken: cancellationToken);
```
Comment (tnotheis) [inline Modules/Messages/src/Messages.Infrastructure/Persistence/Database/Repository/RelationshipsRepository.cs:32]:
```suggestion
            .FirstOrDefaultAsync(cancellationToken);
```

File: Modules/Messages/src/Messages.Infrastructure/Persistence/Database/Repository/RelationshipsRepository.cs [hunk 2]
```diff
@@ -22,4 +23,12 @@ public RelationshipsRepository(MessagesDbContext dbContext)
             .Select(r => r.Id)
             .FirstOrDefaultAsync();
     }
+
+    public Task<Relationship?> FindRelationship(RelationshipId id, CancellationToken cancellationToken, bool track = false)
+    {
+        return _dbContext.Relationships
+            .AsNoTracking()
```
Comment (tnotheis) [inline Modules/Messages/src/Messages.Infrastructure/Persistence/Database/Repository/RelationshipsRepository.cs:30]: Mhh, you always use `AsNoTracking`, no matter what you pass as `track`. Though, when I think about it: we will never ever update a Relationship from the Messages module anyway. So you might as well just remove this parameter.

File: Modules/Relationships/src/Relationships.Application/Relationships/Commands/CreateRelationship/Handler.cs [hunk 3]
```diff
@@ -65,17 +65,29 @@ private async Task CreateAndSaveRelationship()
             _cancellationToken
         );
 
+        var relationships = existingRelationships.ToList();
+
+        EnsureNoTerminatedRelationshipExists(relationships);
```
Comment (tnotheis) [inline Modules/Relationships/src/Relationships.Application/Relationships/Commands/CreateRelationship/Handler.cs:70]: This is domain knowledge. So you should check this in the constructor of the Relationship. Just like all the other checks (e.g. for an existing relationship - which, btw, should just be extended instead of implementing a completely new check)

File: Modules/Relationships/src/Relationships.Application/ApplicationErrors.cs [hunk 4]
```diff
@@ -1,3 +1,30 @@
...
+        {
+            var targetIdentityString = string.IsNullOrEmpty(targetIdentity) ? "the target identity" : targetIdentity;
+
+            return new ApplicationError("error.platform.validation.relationshipRequest.relationshipToTargetAlreadyExists", $"A relationship to {targetIdentityString} already exists.");
+        }
+
+        public static ApplicationError CannotSendRelationshipRequestToYourself()
+        {
+            return new ApplicationError("error.platform.validation.relationshipRequest.cannotSendRelationshipRequestToYourself", "The template you provided is your own. You cannot send a relationship request to yourself.");
+        }
```
Comment (tnotheis) [inline Modules/Relationships/src/Relationships.Application/ApplicationErrors.cs:19]: Why did you add these errors?

File: Modules/Messages/src/Messages.Application/Messages/Commands/SendMessage/Handler.cs [hunk 5]
```diff
@@ -75,6 +75,20 @@ private async Task<List<RecipientInformation>> ValidateRecipients(SendMessageCom
                 throw new OperationFailedException(ApplicationErrors.NoRelationshipToRecipientExists(recipientDto.Address));
             }
 
+            var relationshipBetweenSenderAndRecipient = await _relationshipsRepository.FindRelationship(idOfRelationshipBetweenSenderAndRecipient, cancellationToken);
```
Comment (tnotheis) [inline Modules/Messages/src/Messages.Application/Messages/Commands/SendMessage/Handler.cs:78]: Now there are two database queries executed. One in line 70 (where only the id is fetched) and another one here (where you fetch the whole relationship). This can easily be done in one query.

File: Modules/Relationships/src/Relationships.Application/Relationships/Commands/TerminateRelationship/Handler.cs [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    public async Task<TerminateRelationshipResponse> Handle(TerminateRelationshipCommand request, CancellationToken cancellationToken)
+    {
+        var relationshipId = RelationshipId.Parse(request.RelationshipId);
+        var relationship = await _relationshipsRepository.FindRelationship(relationshipId, _activeIdentity, cancellationToken, track: true);
+
+        relationship.Terminate(_activeIdentity, _activeDevice);
+
+        await _relationshipsRepository.Update(relationship);
+
+        var partnerIdentity = relationship.To == _activeIdentity ? relationship.From : relationship.To;
```
Comment (tnotheis) [inline Modules/Relationships/src/Relationships.Application/Relationships/Commands/TerminateRelationship/Handler.cs:35]: We usually call the relationship partner `peer`.

File: Modules/Relationships/src/Relationships.Application/Relationships/Commands/TerminateRelationship/Handler.cs [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+        _eventBus.Publish(new RelationshipStatusChangedIntegrationEvent(relationship));
+        _eventBus.Publish(new RelationshipTerminatedIntegrationEvent(relationship, partnerIdentity));
```
Comment (tnotheis) [inline Modules/Relationships/src/Relationships.Application/Relationships/Commands/TerminateRelationship/Handler.cs:38]: Is it necessary to publish both integration events? Isn't the first one sufficient?

File: Modules/Relationships/test/Relationships.Application.Tests/Tests/Relationships/Commands/TerminateRelationship/HandlerTests.cs [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+        // Assert
+        A.CallTo(
+                () => mockRelationshipsRepository.Update(
+                    A<Relationship>.That.Matches(r => r.Id == relationship.Id && r.Status == RelationshipStatus.Terminated))
+            )
+            .MustHaveHappenedOnceExactly();
+    }
+
+    [Fact]
+    public async Task Publishes_RelationshipStatusChangedIntegrationEvent()
```
Comment (tnotheis) [inline Modules/Relationships/test/Relationships.Application.Tests/Tests/Relationships/Commands/TerminateRelationship/HandlerTests.cs:81]: IF the `RelationshipTerminatedIntegrationEvent` is actually required, you should test that it is published.

File: Modules/Synchronization/src/Synchronization.Application/IntegrationEvents/Incoming/RelationshipTerminated/RelationshipTerminatedIntegrationEvent.cs
Comment (tnotheis) [inline Modules/Synchronization/src/Synchronization.Application/IntegrationEvents/Incoming/RelationshipTerminated/RelationshipTerminatedIntegrationEvent.cs:1]: Remember to remove this if you remove the sending side

**Review by tnotheis — CHANGES_REQUESTED** [review]

File: Modules/Messages/src/Messages.Application/ApplicationErrors.cs [hunk 9]
```diff
@@ -17,4 +17,11 @@ public static ApplicationError MaxNumberOfUnreceivedMessagesReached(string recip
 
         return new ApplicationError("error.platform.validation.message.maxNumberOfUnreceivedMessagesReached", $"The message could not be sent because {recipientText} already has the maximum number of unread messages from you.");
     }
+
+    public static ApplicationError RelationshipToRecipientNotActive(string recipient = "")
+    {
+        var recipientText = string.IsNullOrEmpty(recipient) ? "one of the recipients" : recipient;
+
+        return new ApplicationError("error.platform.validation.message.relationshipToRecipientNotActive", $"Cannot send message to {recipientText} because the relationship to it is not active.");
```
[the remark this answers is not in the document]
Comment (NikolaVetnic) [reply Modules/Messages/src/Messages.Application/ApplicationErrors.cs:25]: Nice, good idea. Changes made.

File: Modules/Messages/src/Messages.Application/Messages/Commands/SendMessage/Handler.cs [hunk 10]
```diff
@@ -75,6 +75,20 @@ private async Task<List<RecipientInformation>> ValidateRecipients(SendMessageCom
...
+            {
+                _logger.LogInformation("Sending message aborted. There is no relationship between sender ({sender}) and recipient ({recipient}).", sender, recipientDto.Address);
+                throw new OperationFailedException(ApplicationErrors.NoRelationshipToRecipientExists(recipientDto.Address));
+            }
+
+            if (relationshipBetweenSenderAndRecipient.Status != RelationshipStatus.Active)
+            {
+                _logger.LogInformation("Sending message aborted. Relationship between sender ({sender}) and recipient ({recipient}) is not active.", sender, recipientDto.Address);
+                throw new OperationFailedException(ApplicationErrors.RelationshipToRecipientNotActive(recipientDto.Address));
+            }
```
[the remark this answers is not in the document]
Comment (NikolaVetnic) [reply Modules/Messages/src/Messages.Application/Messages/Commands/SendMessage/Handler.cs:90]: Changes made.

File: Modules/Messages/src/Messages.Infrastructure/Persistence/Database/Repository/RelationshipsRepository.cs (hunk 1 above)
Comment (NikolaVetnic) [reply Modules/Messages/src/Messages.Infrastructure/Persistence/Database/Repository/RelationshipsRepository.cs:32]: Changes made.

File: Modules/Messages/src/Messages.Infrastructure/Persistence/Database/Repository/RelationshipsRepository.cs (hunk 2 above)
Comment (NikolaVetnic) [reply Modules/Messages/src/Messages.Infrastructure/Persistence/Database/Repository/RelationshipsRepository.cs:30]: Probably because I copied the signature from another place, I agree that no tracking is required for relationships from the Messages module. 

Changes made.

File: Modules/Relationships/src/Relationships.Application/ApplicationErrors.cs (hunk 4 above)
Comment (NikolaVetnic) [reply Modules/Relationships/src/Relationships.Application/ApplicationErrors.cs:19]: I've used them at some point and then I refactored the code. I forgot to remove these unfortunately.

Changes made.

File: Modules/Relationships/src/Relationships.Application/Relationships/Commands/CreateRelationship/Handler.cs (hunk 3 above)
Comment (tnotheis) [reply Modules/Relationships/src/Relationships.Application/Relationships/Commands/CreateRelationship/Handler.cs:70]: I think I like `EnsureNotStatus`. Even though the name sucks, but I can't find something better either. ;)

File: Modules/Messages/src/Messages.Application/Messages/Commands/SendMessage/Handler.cs (hunk 5 above)
Comment (NikolaVetnic) [reply Modules/Messages/src/Messages.Application/Messages/Commands/SendMessage/Handler.cs:78]: Yes, sorry.

Changes made.

File: Modules/Relationships/src/Relationships.Application/Relationships/Commands/TerminateRelationship/Handler.cs (hunk 6 above)
Comment (NikolaVetnic) [reply Modules/Relationships/src/Relationships.Application/Relationships/Commands/TerminateRelationship/Handler.cs:35]: Changes made.

File: Modules/Relationships/src/Relationships.Application/Relationships/Commands/TerminateRelationship/Handler.cs (hunk 7 above)
Comment (NikolaVetnic) [reply Modules/Relationships/src/Relationships.Application/Relationships/Commands/TerminateRelationship/Handler.cs:38]: I published both for the sake of completeness but of course, the second one suffices.

Changes made.

File: Modules/Relationships/test/Relationships.Application.Tests/Tests/Relationships/Commands/TerminateRelationship/HandlerTests.cs (hunk 8 above)
Comment (NikolaVetnic) [reply Modules/Relationships/test/Relationships.Application.Tests/Tests/Relationships/Commands/TerminateRelationship/HandlerTests.cs:81]: Changes made.

File: Modules/Relationships/src/Relationships.Application/Relationships/Commands/CreateRelationship/Handler.cs (hunk 3 above)
Comment (NikolaVetnic) [reply Modules/Relationships/src/Relationships.Application/Relationships/Commands/CreateRelationship/Handler.cs]: Yes, the name sucks aesthetically but it works :)

Changes made.

File: Modules/Relationships/src/Relationships.Application/Relationships/Commands/TerminateRelationship/Handler.cs (hunk 7 above)
Comment (NikolaVetnic) [reply Modules/Relationships/src/Relationships.Application/Relationships/Commands/TerminateRelationship/Handler.cs]: Ok, `RelationshipTerminatedIntegrationEvent` removed. 

Changes made.

File: Modules/Messages/src/Messages.Domain/Entities/Relationship.cs [hunk 11]
```diff
@@ -24,6 +25,15 @@ private Relationship()
     public DateTime CreatedAt { get; }
 
     public RelationshipStatus Status { get; }
+
+    public void EnsureSendingMessagesIsAllowed(int numberOfUnreceivedMessagesFromActiveIdentity, int maxNumberOfUnreceivedMessagesFromOneSender)
```
Comment (tnotheis) [inline Modules/Messages/src/Messages.Domain/Entities/Relationship.cs:29]: Tests for this method are missing

**Review by tnotheis — CHANGES_REQUESTED** [review]

File: Modules/Relationships/src/Relationships.Application/Relationships/Commands/CreateRelationship/Handler.cs [hunk 12]
```diff
@@ -67,7 +67,8 @@ private async Task CreateAndSaveRelationship()
 
         var relationships = existingRelationships.ToList();
 
-        EnsureNoTerminatedRelationshipExists(relationships);
+        foreach (var relationship in relationships)
+            relationship.EnsureNotStatus(RelationshipStatus.Terminated);
```
[the remark this answers is not in the document]
Comment (tnotheis) [reply Modules/Relationships/src/Relationships.Application/Relationships/Commands/CreateRelationship/Handler.cs:71]: And please remember to add corresponding tests for the constructor then.

File: Modules/Messages/src/Messages.Application/Messages/Commands/SendMessage/Handler.cs [hunk 13]
```diff
@@ -67,40 +67,22 @@ private async Task<List<RecipientInformation>> ValidateRecipients(SendMessageCom
...
             }
 
-            if (relationshipBetweenSenderAndRecipient.Status != RelationshipStatus.Active)
-            {
-                _logger.LogInformation("Sending message aborted. Relationship between sender ({sender}) and recipient ({recipient}) is not active.", sender, recipientDto.Address);
-                throw new OperationFailedException(ApplicationErrors.RelationshipToRecipientNotActive(recipientDto.Address));
-            }
-
             var numberOfUnreceivedMessagesFromActiveIdentity = await _messagesRepository.CountUnreceivedMessagesFromSenderToRecipient(sender, recipientDto.Address, cancellationToken);
+            var maxNumberOfUnreceivedMessagesFromOneSender = _options.MaxNumberOfUnreceivedMessagesFromOneSender;
```
[the remark this answers is not in the document]
Comment (NikolaVetnic) [reply Modules/Messages/src/Messages.Application/Messages/Commands/SendMessage/Handler.cs:79]: Hehe, no prob, changes made.

File: Modules/Relationships/src/Relationships.Application/IntegrationEvents/Outgoing/RelationshipTerminatedIntegrationEvent.cs
[the remark this answers is not in the document]
Comment (NikolaVetnic) [reply Modules/Relationships/src/Relationships.Application/IntegrationEvents/Outgoing/RelationshipTerminatedIntegrationEvent.cs:1]: Yes, it reported "1 reference", but that was the constructor. It looked as if though it was used.

Changes made.

File: Modules/Relationships/src/Relationships.Application/Relationships/Commands/CreateRelationship/Handler.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (NikolaVetnic) [reply Modules/Relationships/src/Relationships.Application/Relationships/Commands/CreateRelationship/Handler.cs:71]: Ok, I've removed the `NotInStatus` method then and added this so that it corresponds to the other `Ensure` we have:
```csharp
    private static void EnsureNoRelationshipInStatus(RelationshipStatus status, List<Relationship> existingRelationships)
    {
        if (existingRelationships.Any(r => r.Status == status))
            throw new DomainException(DomainErrors.RelationshipIsInIncorrectStatus(status));
    }
```

Changes made.

File: Modules/Relationships/src/Relationships.Application/Relationships/Commands/TerminateRelationship/Handler.cs [hunk 14]
```diff
@@ -31,11 +30,9 @@ public async Task<TerminateRelationshipResponse> Handle(TerminateRelationshipCom
         relationship.Terminate(_activeIdentity, _activeDevice);
 
         await _relationshipsRepository.Update(relationship);
-
-        var partnerIdentity = relationship.To == _activeIdentity ? relationship.From : relationship.To;
+        _ = relationship.To == _activeIdentity ? relationship.From : relationship.To;
```
[the remark this answers is not in the document]
Comment (NikolaVetnic) [reply Modules/Relationships/src/Relationships.Application/Relationships/Commands/TerminateRelationship/Handler.cs:33]: Wtf indeed, I had no idea something like this could happen... This is how it was at first:
```csharp
        var peer = relationship.To == _activeIdentity ? relationship.From : relationship.To;

        _eventBus.Publish(new RelationshipStatusChangedIntegrationEvent(relationship));
        _eventBus.Publish(new RelationshipTerminatedIntegrationEvent(relationship, peer));
```

Then I removed the second event:
```csharp
        var peer = relationship.To == _activeIdentity ? relationship.From : relationship.To;

        _eventBus.Publish(new RelationshipStatusChangedIntegrationEvent(relationship));
```

And then I ran `dotnet format` to fix the formatting:
```csharp
        _ = relationship.To == _activeIdentity ? relationship.From : relationship.To;

        _eventBus.Publish(new RelationshipStatusChangedIntegrationEvent(relationship));
```

I had no idea `dotnet format` can make such radical changes in code.

Changes made.

File: Modules/Relationships/src/Relationships.Domain/DomainErrors.cs [hunk 15]
```diff
@@ -35,8 +35,15 @@ public static DomainError RelationshipIsNotInCorrectStatus(RelationshipStatus ex
             $"The relationship has to be in status '{expectedStatus}' to perform this action.");
     }
 
+    public static DomainError RelationshipIsInIncorrectStatus(RelationshipStatus expectedStatus)
+    {
+        return new DomainError("error.platform.validation.relationshipRequest.relationshipIsInIncorrectStatus",
+            $"The relationship cannot to be in status '{expectedStatus}' to perform this action.");
```
[the remark this answers is not in the document]
Comment (NikolaVetnic) [reply Modules/Relationships/src/Relationships.Domain/DomainErrors.cs:41]: Sure, changes made.

File: Modules/Relationships/src/Relationships.Domain/Aggregates/Relationships/Relationship.cs [hunk 16]
```diff
@@ -157,6 +157,12 @@ public void Terminate(IdentityAddress activeIdentity, DeviceId activeDevice)
         AuditLog.Add(auditLogEntry);
     }
 
+    public void EnsureNotStatus(RelationshipStatus status)
```
[the remark this answers is not in the document]
Comment (NikolaVetnic) [reply Modules/Relationships/src/Relationships.Domain/Aggregates/Relationships/Relationship.cs:160]: Well... I wouldn't say I agree with you 100% on this one, but I think it's a matter of taste, not something that affects functionality.

Changes made.

File: Modules/Messages/src/Messages.Domain/Entities/Relationship.cs (hunk 11 above)
Comment (NikolaVetnic) [reply Modules/Messages/src/Messages.Domain/Entities/Relationship.cs:29]: Should I do a simple domain test for the method in isolation, or should I test the entire handler?

Comment (NikolaVetnic) [reply Modules/Messages/src/Messages.Domain/Entities/Relationship.cs:29]: Changes made.

File: Modules/Messages/test/Messages.Domain.Tests/Relationships/RelationshipTests.cs [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+public class RelationshipTests
+{
+    [Fact]
+    public void Relationship_must_be_active_to_allow_sending_messages()
+    {
+        var relationship = CreateRelationship(RelationshipStatus.Pending);
+
+        var acting = () => relationship.EnsureSendingMessagesIsAllowed(5, 5);
+
+        //relationship.Should().NotBe(null);
```
Comment (tnotheis) [inline Modules/Messages/test/Messages.Domain.Tests/Relationships/RelationshipTests.cs:20]: ?

File: Modules/Messages/test/Messages.Domain.Tests/Relationships/RelationshipTests.cs [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    [Fact]
+    public void Max_number_of_unrecieved_messages_cannot_be_reached_to_allow_sending_messages()
+    {
+        var relationship = CreateRelationship();
+
+        var acting = () => relationship.EnsureSendingMessagesIsAllowed(5, 5);
+
+        //relationship.Should().NotBe(null);
```
Comment (tnotheis) [inline Modules/Messages/test/Messages.Domain.Tests/Relationships/RelationshipTests.cs:31]: ?

File: Modules/Messages/test/Messages.Domain.Tests/Relationships/RelationshipTests.cs [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+        var relationship = CreateRelationship(RelationshipStatus.Pending);
+
+        var acting = () => relationship.EnsureSendingMessagesIsAllowed(5, 5);
+
+        //relationship.Should().NotBe(null);
+        acting.Should().Throw<DomainException>().Which.Code.Should().Be("error.platform.validation.message.relationshipToRecipientNotActive");
+    }
+
+    [Fact]
+    public void Max_number_of_unrecieved_messages_cannot_be_reached_to_allow_sending_messages()
```
Comment (tnotheis) [inline Modules/Messages/test/Messages.Domain.Tests/Relationships/RelationshipTests.cs:25]:
```suggestion
    public void Max_number_of_unreceived_messages_cannot_be_reached_to_allow_sending_messages()
```

Comment (tnotheis) [inline Modules/Messages/test/Messages.Domain.Tests/Relationships/RelationshipTests.cs:25]:
```suggestion
    public void Max_number_of_unreceived_messages_must_not_be_reached()
```

File: Modules/Messages/test/Messages.Domain.Tests/Relationships/RelationshipTests.cs [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+using Backbone.Modules.Messages.Domain.Ids;
+using Backbone.UnitTestTools.Data;
+using FluentAssertions;
+using Xunit;
+
+namespace Backbone.Modules.Messages.Domain.Tests.Relationships;
+public class RelationshipTests
+{
+    [Fact]
+    public void Relationship_must_be_active_to_allow_sending_messages()
```
Comment (tnotheis) [inline Modules/Messages/test/Messages.Domain.Tests/Relationships/RelationshipTests.cs:14]: Add another test to make sure it's not allowed to send messages in status `Terminated`.

File: Modules/Relationships/src/Relationships.Domain/Aggregates/Relationships/Relationship.cs [hunk 21]
```diff
@@ -25,6 +25,9 @@ public Relationship(RelationshipTemplate relationshipTemplate, IdentityAddress a
         EnsureTargetIsNotSelf(relationshipTemplate, activeIdentity);
         EnsureNoActiveRelationshipToTargetExists(relationshipTemplate.CreatedBy, existingRelationships);
 
+        var listOfAllowedStatuses = new List<RelationshipStatus> { RelationshipStatus.Pending, RelationshipStatus.Active, RelationshipStatus.Rejected, RelationshipStatus.Revoked };
```
Comment (tnotheis) [inline Modules/Relationships/src/Relationships.Domain/Aggregates/Relationships/Relationship.cs:28]: is it just me, or is "allowedStatuses" a confusing/misleading name? 
1. All statuses are "allowed".
2. If you mean "allowed" in the sense of "existing relationships can have these statuses in order to be able to create a new relationship", then the list has to be inversed, doesn't it? Because if there is e.g. an `Active` relationship, we should not be allowed to create another one.

File: Modules/Relationships/src/Relationships.Domain/DomainErrors.cs [hunk 22]
```diff
@@ -35,10 +35,10 @@ public static DomainError RelationshipIsNotInCorrectStatus(RelationshipStatus ex
             $"The relationship has to be in status '{expectedStatus}' to perform this action.");
     }
 
-    public static DomainError RelationshipIsInIncorrectStatus(RelationshipStatus expectedStatus)
+    public static DomainError RelationshipIsInIncorrectStatus(RelationshipStatus undesiredStatus)
     {
         return new DomainError("error.platform.validation.relationshipRequest.relationshipIsInIncorrectStatus",
-            $"The relationship cannot to be in status '{expectedStatus}' to perform this action.");
+            $"The relationship cannot to be in status '{undesiredStatus}' to perform this action.");
     }
```
Comment (tnotheis) [inline Modules/Relationships/src/Relationships.Domain/DomainErrors.cs:42]: As described in a comment in my previous review, I would change the error code to something like "*HasToBeInStatus" (or however we call it in other places), and in the error message list all statuses in which the operation is possible.

File: Modules/Relationships/test/Relationships.Domain.Tests/Tests/Aggregates/Relationships/CreateRelationshipTests.cs [hunk 23]
```diff
@@ -95,4 +95,20 @@ public void Creating_a_new_relationship_is_possible_if_existing_ones_are_rejecte
         // Assert
         acting.Should().NotThrow<DomainException>();
     }
+
```
Comment (tnotheis) [inline Modules/Relationships/test/Relationships.Domain.Tests/Tests/Aggregates/Relationships/CreateRelationshipTests.cs:98]: I just noticed that in line 88 there is a commented out line. Probably my bad. Can you uncomment it please?

File: Modules/Messages/test/Messages.Domain.Tests/Relationships/RelationshipTests.cs [hunk 24]
```diff
@@ -0,0 +1,70 @@
...
+        from ??= TestDataGenerator.CreateRandomIdentityAddress();
+        to ??= TestDataGenerator.CreateRandomIdentityAddress();
+        createdAt ??= DateTime.UtcNow;
+        status ??= RelationshipStatus.Active;
+
+        SetBackingField(relationship, "Id", RelationshipId.Parse(relationshipId));
+        SetBackingField(relationship, "From", from);
+        SetBackingField(relationship, "To", to);
+        SetBackingField(relationship, "CreatedAt", createdAt);
+        SetBackingField(relationship, "Status", status);
```
Comment (tnotheis) [inline Modules/Messages/test/Messages.Domain.Tests/Relationships/RelationshipTests.cs:58]:
```c#
    public static Relationship LoadForTesting(...)
    {
    }
```

**Review by tnotheis — CHANGES_REQUESTED** [review]

File: Modules/Relationships/test/Relationships.Domain.Tests/Tests/Aggregates/Relationships/CreateRelationshipTests.cs (hunk 23 above)
Comment (NikolaVetnic) [reply Modules/Relationships/test/Relationships.Domain.Tests/Tests/Aggregates/Relationships/CreateRelationshipTests.cs:98]: Sure, I left it because I thought you did it on purpose.

Changes made.

File: Modules/Messages/test/Messages.Domain.Tests/Relationships/RelationshipTests.cs (hunk 17 above)
Comment (NikolaVetnic) [reply Modules/Messages/test/Messages.Domain.Tests/Relationships/RelationshipTests.cs:20]: Changes made.

File: Modules/Messages/test/Messages.Domain.Tests/Relationships/RelationshipTests.cs (hunk 18 above)
Comment (NikolaVetnic) [reply Modules/Messages/test/Messages.Domain.Tests/Relationships/RelationshipTests.cs:31]: Changes made.

File: Modules/Messages/test/Messages.Domain.Tests/Relationships/RelationshipTests.cs (hunk 19 above)
Comment (NikolaVetnic) [reply Modules/Messages/test/Messages.Domain.Tests/Relationships/RelationshipTests.cs:25]: Changes made.

Comment (NikolaVetnic) [reply Modules/Messages/test/Messages.Domain.Tests/Relationships/RelationshipTests.cs:25]: Changes made.

File: Modules/Messages/test/Messages.Domain.Tests/Relationships/RelationshipTests.cs (hunk 20 above)
Comment (NikolaVetnic) [reply Modules/Messages/test/Messages.Domain.Tests/Relationships/RelationshipTests.cs:14]: I did this:
```c#
    [Fact]
    public void Relationship_cannot_be_terminated_to_allow_sending_messages()
    {
        // Arrange
        var relationship = CreateRelationship(RelationshipStatus.Terminated);

        // Act
        var acting = () => relationship.EnsureSendingMessagesIsAllowed(0, 5);

        // Assert
        acting.Should().Throw<DomainException>().Which.Code.Should().Be("error.platform.validation.message.relationshipToRecipientNotActive");
    }
```

It is very similar to the first test in the file though. Maybe I misunderstood you.

Changes made.

File: Modules/Relationships/src/Relationships.Domain/Aggregates/Relationships/Relationship.cs (hunk 21 above)
Comment (NikolaVetnic) [reply Modules/Relationships/src/Relationships.Domain/Aggregates/Relationships/Relationship.cs:28]: Ok, I've changed the variable name to `listOfStatusesRelationshipCanBeInWhenCreatingNewRelationship`... I know it's long but at least it's unambiguous.

Changes made.

File: Modules/Messages/test/Messages.Domain.Tests/Relationships/RelationshipTests.cs (hunk 20 above)
Comment (tnotheis) [reply Modules/Messages/test/Messages.Domain.Tests/Relationships/RelationshipTests.cs:14]: Yeah. Just one minor comment regarding the phrasing: maybe the tests should be called "Sending_messages_is_not_allowed_when_relationship_is_*"

Comment (tnotheis) [reply Modules/Messages/test/Messages.Domain.Tests/Relationships/RelationshipTests.cs:14]: Ahm, and correspondingly "Sending_messages_is_allowed_when_relationship_is_active"

File: Modules/Relationships/src/Relationships.Domain/Aggregates/Relationships/Relationship.cs (hunk 21 above)
Comment (tnotheis) [reply Modules/Relationships/src/Relationships.Domain/Aggregates/Relationships/Relationship.cs:28]: That's not accurate. It only mentions one relationship. So it sounds as if you mean the relationship that is about to be created. 

Maybe we shouldn't create a variable, and just use it inline. So:


```c#
![RelationshipStatus.Pending, RelationshpStatus.Active, ...].Contains(r.Status)
```

File: Modules/Relationships/test/Relationships.Domain.Tests/Tests/Aggregates/Relationships/CreateRelationshipTests.cs (hunk 23 above)
Comment (NikolaVetnic) [reply Modules/Relationships/test/Relationships.Domain.Tests/Tests/Aggregates/Relationships/CreateRelationshipTests.cs:97]: Sure, done.

Changes made.

File: Modules/Messages/test/Messages.Domain.Tests/Relationships/RelationshipTests.cs (hunk 24 above)
Comment (NikolaVetnic) [reply Modules/Messages/test/Messages.Domain.Tests/Relationships/RelationshipTests.cs]: Changes made.

File: Modules/Relationships/test/Relationships.Domain.Tests/Tests/Aggregates/Relationships/CreateRelationshipTests.cs (hunk 23 above)
Comment (tnotheis) [reply Modules/Relationships/test/Relationships.Domain.Tests/Tests/Aggregates/Relationships/CreateRelationshipTests.cs:97]: You weren't suppose to remove the commented out line. Just to uncomment it.

File: Modules/Relationships/src/Relationships.Domain/Aggregates/Relationships/Relationship.cs (hunk 21 above)
Comment (tnotheis) [reply Modules/Relationships/src/Relationships.Domain/Aggregates/Relationships/Relationship.cs]: Ahh, I just saw that this isn't possible. :( I will think about it tomorrow.

File: Modules/Relationships/src/Relationships.Domain/Aggregates/Relationships/Relationship.cs [hunk 25]
```diff
@@ -25,8 +25,8 @@ public Relationship(RelationshipTemplate relationshipTemplate, IdentityAddress a
         EnsureTargetIsNotSelf(relationshipTemplate, activeIdentity);
         EnsureNoActiveRelationshipToTargetExists(relationshipTemplate.CreatedBy, existingRelationships);
 
-        var listOfAllowedStatuses = new List<RelationshipStatus> { RelationshipStatus.Pending, RelationshipStatus.Active, RelationshipStatus.Rejected, RelationshipStatus.Revoked };
-        EnsureStatus(listOfAllowedStatuses, existingRelationships);
+        var listOfStatusesRelationshipCanBeInWhenCreatingNewRelationship = new List<RelationshipStatus> { RelationshipStatus.Rejected, RelationshipStatus.Revoked };
```
Comment (tnotheis) [inline Modules/Relationships/src/Relationships.Domain/Aggregates/Relationships/Relationship.cs:28]: Reminder for myself: find a better name

**Review by tnotheis — CHANGES_REQUESTED** [review]
